clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreBufferService.java

```diff
@@ -340,13 +343,31 @@ public void run() {
   private final List<StoreBufferDrainer> drainerList = new ArrayList<>();
   private final long bufferCapacityPerDrainer;
 
-  public StoreBufferService(int drainerNum, long bufferCapacityPerDrainer, long bufferNotifyDelta) {
+  private final RecordHandler leaderRecordHandler;
+
+  public StoreBufferService(
+      int drainerNum,
+      long bufferCapacityPerDrainer,
+      long bufferNotifyDelta,
+      boolean queueLeaderWrites) {
     this.drainerNum = drainerNum;
     this.blockingQueueArr = new ArrayList<>();
     this.bufferCapacityPerDrainer = bufferCapacityPerDrainer;
     for (int cur = 0; cur < drainerNum; ++cur) {
       this.blockingQueueArr.add(new MemoryBoundBlockingQueue<>(bufferCapacityPerDrainer, bufferNotifyDelta));
     }
+    this.leaderRecordHandler = queueLeaderWrites ? this::queueLeaderRecord : this::processLeaderRecord;
```

Comment:
Have a few questions regarding this config
1. Is the producer callback queue tunable? If I read the codes right, the queue size is equal to the number of drainer threads? They used to be a queue but now just a single thread.
2. Under which circumstances, this flag should be turn off? I suppose the most of cases, we don't?

Reply:
Thanks Adam,

1. The producer callback queue is not tunable AFAIK, so this might become a bottleneck. That is the reason for making the new mode an opt-in config, so we can test it carefully.

2. Depending on the result of the above test, we may decide that: a) this is a good default mode, and change it to opt-out, b) that it's not good at all, and that we should just delete it, c) that it's not good yet but might still be good later (e.g. in conjunction with the shared producer mode, perhaps), or d) ...?